SILOptimizer: allow function pointers in static globals and static global arrays.

eeckstein opened the pull request:
For example, generate a static global for the array literal in:
```
func foo(_ i: Int) -> Int { ... }
func bar(_ i: Int) -> Int { ... }

func returnFunctionArray() -> [(Int) -> Int] {
  return [foo, bar]
}
```
rdar://73570149
https://bugs.swift.org/browse/SR-14101

This PR also contains an improvement of DeadFunctionElimination: it now can also remove dead global variables.
Handling function references in global initializers is required in DeadFunctionElimination anyway. It was only a small addition to implement dead global variable elimination in the same change.

rdar://32956923

eeckstein committed:
SIL: add FunctionRefInst::getReferencedFunction()

If we know that we have a FunctionRefInst (and not another variant of FunctionRefBaseInst), we know that getting the referenced function will not be null (in contrast to FunctionRefBaseInst::getReferencedFunctionOrNull).

NFC

eeckstein committed:
SIL: allow creating convert_function and thin_to_thick_function instructions in an initializer of a SILGlobalVariable.

This means: don't require to have a SILFunction in the SILBuilder.

eeckstein committed:
IRGen: a small refactoring in GenConstant

Just moving a few functions around and make emitConstantValue the main entry point for creating constants.

NFC

eeckstein committed:
SILOptimizer: allow function pointers in static globals and static global arrays.

For example, generate a static global for the array literal in:

func foo(_ i: Int) -> Int { ... }
func bar(_ i: Int) -> Int { ... }

func returnFunctionArray() -> [(Int) -> Int] {
  return [foo, bar]
}

rdar://73570149
https://bugs.swift.org/browse/SR-14101

aschwaighofer commented on include/swift/SIL/SILInstruction.h:
```diff
@@ -2816,6 +2816,9 @@ class FunctionRefInst : public FunctionRefBaseInst {
                   TypeExpansionContext context);
 
 public:
+  /// Return the referenced function.
```
I had intentionally removed `getReferencedFunction` and replaced it with `getInitiallyReferencedFunction` in `FunctionRefBaseInst` to make clear to clients that the referenced function returned might not actually what is dynamically called (e.g. in case of `DynamicFunctionRefInst`)

Why do you need to reintroduce it? The presence of `getReferencedFunction` makes the distinction murky.


```
  /// Return the referenced function if this is a function_ref instruction and  
  /// therefore a client can rely on the dynamically called function being equal
  /// to the returned value and null otherwise.                                 
  SILFunction *getReferencedFunctionOrNull() const {                            
    auto kind = getKind();                                                      
    if (kind == SILInstructionKind::FunctionRefInst)                            
      return f;                                                                 
    assert(kind == SILInstructionKind::DynamicFunctionRefInst ||                
           kind == SILInstructionKind::PreviousDynamicFunctionRefInst);         
    return nullptr;                                                             
  }                                                                             
                                                                                
  /// Return the initially referenced function.                                 
  ///                                                                           
  /// WARNING: This not necessarily the function that will be called at runtime.
  /// If the callee is a (prev_)dynamic_function_ref the actual function called 
  /// might be different because it could be dynamically replaced at runtime.   
  ///                                                                           
  /// If the client of this API wants to look at the content of the returned SIL
  /// function it should call getReferencedFunctionOrNull() instead.            
  SILFunction *getInitiallyReferencedFunction() const { return f; }   
```

eeckstein replied:
But this is not the case for the derived FunctionRefInst, right?
Not having it on the base class makes sense, but why not in FunctionRefInst?

eeckstein replied:
The reason why I re-introduced it was that both names, `getReferencedFunctionOrNull` and `getInitiallyReferencedFunction()` don't really make sense for the "static" `function_ref`

aschwaighofer replied:
I mis-parsed the line `class FunctionRefInst : public FunctionRefBaseInst {` even though I had read your commit message (which also says `FunctionRefInst`) I was under the impression this added it to the base.

Ignore me. This is of course fine.

aschwaighofer reviewed (approved):
LGTM

eeckstein committed:
DeadFunctionElimination: Also eliminate dead global variables.

It makes sense to to this in a single pass, because there might be dead cycles for globals and functions, e.g. if a global references a function in its static initializer and the function references that global.

Another improvement: eliminate dead global-initializers. Before we had an explicit SIL representation of statically initialized globals, we had to keep them alive.

rdar://32956923

eeckstein committed:
DeadFunctionElimination: rename the pass and some other refactoring

The main change is to rename DeadFunctionElimination -> DeadFunctionAndGlobalElimination, because the pass is now also doing dead-global elimination.
A second change is to remove the FunctionLivenessComputation base class. It’s not used anywhere else.

eeckstein commented:
@swift-ci test

eeckstein commented:
@swift-ci test

eeckstein commented:
@swift-ci benchmark

eeckstein commented:
@swift-ci test linux

eeckstein commented:
@swift-ci test macOS

eeckstein committed:
SILOptimizer: prevent illegal load forwarding of function references in global variables.

We cannot replace a load from a global let-variable with a function_ref, if the referenced function would violate the resilience rules.
That means if a non-public function_ref would be inlined into a function which is serialized.

eeckstein commented:
@swift-ci test

eeckstein commented:
@swift-ci test
